### PR TITLE
Add folding code for CollectivePermuteOp

### LINKED ISSUE #2406: Optimize collective_permute source_target_pairs size == 0 lowering
Similar to this [issue](https://github.com/tenstorrent/tt-mlir/issues/2223) in all_gather where we replace replica_groups size = 1 with a noop, for the case of collective_permute, when we get a source_target_pairs with size = 0, we need to also convert that into a nop, as that means no tensor data is moving from any device, and converting it into a noop won't change the semantic meaning of the graph.

This is blocked by this PR: https://github.com/tenstorrent/tt-mlir/pull/2405

### PR DESCRIPTION
**Pull request by hkwonTT**
### Ticket
https://github.com/tenstorrent/tt-mlir/issues/2406
closes #2406 

### Problem description
CollectivePermute Op is semantically meaningless when there is no tensor data movement from any device.
source_target_pair attribute describes where Tensor data moving from and to. There will be no data movement If source and target are same. Also if we don't have any effective movement description, we don't need to execute this op.

### What's changed
- Remove entries in the permute attribute where 'src' and 'target' are identical.
- Exclude the CollectivePermuteOp if the source_target_pair array becomes empty after cleanup.
- Added unit tests to verify folding

### Checklist
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: lib/Dialect/TTNN/IR/TTNNOps.cpp [hunk 1]
```diff
@@ -1747,6 +1747,41 @@ ::mlir::LogicalResult CollectivePermuteOp::verify() {
...
+      auto newShape = llvm::ArrayRef<int64_t>(shape);
+      auto newType = RankedTensorType::get(
+          newShape, srcTargetPairs.getType().getElementType());
+      setSourceTargetPairsAttr(
+          DenseIntElementsAttr::get(newType, filteredPairs));
+    }
+  } else {
+    // No permutations left. Exclude this op.
+    return getInput();
+  }
```
Comment (azecevicTT) [inline lib/Dialect/TTNN/IR/TTNNOps.cpp:1782]: You can reorganize this branch a bit
```cpp
if (filteredPairs.empty()) {
  return getInput();
}
if (srcTargetPairs.getNumElements() !=
        static_cast<int64_t>(filteredPairs.size())) {
  ...
}
return {};

File: lib/Dialect/TTNN/IR/TTNNOps.cpp [hunk 2]
```diff
@@ -1747,6 +1747,41 @@ ::mlir::LogicalResult CollectivePermuteOp::verify() {
...
+
+  if (filteredPairs.size() != 0) {
+    // There are effective permutations left.
+    if (srcTargetPairs.getNumElements() !=
+        static_cast<int64_t>(filteredPairs.size())) {
+      // Update source_target_pairs if changed.
+      std::array<int64_t, 2> shape = {
+          static_cast<int64_t>(filteredPairs.size() / 2), 2};
+      auto newShape = llvm::ArrayRef<int64_t>(shape);
```
Comment (azecevicTT) [inline lib/Dialect/TTNN/IR/TTNNOps.cpp:1773]: You rarely need to construct `ArrayRef` explicitly, you can use `shape` directly in place of `newShape` in `RankedTensorType::get(...)`.

File: lib/Dialect/TTNN/IR/TTNNOps.cpp [hunk 3]
```diff
@@ -1747,6 +1747,41 @@ ::mlir::LogicalResult CollectivePermuteOp::verify() {
   return success();
 }
 
+::mlir::OpFoldResult
+mlir::tt::ttnn::CollectivePermuteOp::fold(FoldAdaptor adaptor) {
+  ::mlir::DenseIntElementsAttr srcTargetPairs = getSourceTargetPairs();
+
+  // Filter out self-mapping src-target pairs
```
Comment (azecevicTT) [inline lib/Dialect/TTNN/IR/TTNNOps.cpp:1754]: Nit: `.` at the end of the comment.

File: lib/Dialect/TTNN/IR/TTNNOps.cpp [hunk 4]
```diff
@@ -1747,6 +1747,41 @@ ::mlir::LogicalResult CollectivePermuteOp::verify() {
...
+  SmallVector<APInt> filteredPairs;
+  for (auto it = srcTargetPairs.begin(), e = srcTargetPairs.end(); it != e;) {
+    auto src = *it++;
+    auto target = *it++;
+    if (src == target) {
+      continue;
+    }
+    filteredPairs.push_back(src);
+    filteredPairs.push_back(target);
+  }
```
Comment (azecevicTT) [inline lib/Dialect/TTNN/IR/TTNNOps.cpp:1764]: I think you can rewrite this without iterators
```suggestion
  auto elements = srcTargetPairs.getValues<APInt>();
  SmallVector<APInt> filteredPairs;
  for (size_t i = 0; i < elements.size(); i += 2) {
    auto src = elements[i];
    auto target = elements[i + 1];
    if (src == target) {
      continue;
    }
    filteredPairs.push_back(src);
    filteredPairs.push_back(target);
  }
```

File: test/ttmlir/Dialect/TTNN/ccl/collective_permute/collective_permute_positive.mlir [hunk 5]
```diff
@@ -1,14 +1,54 @@
...
+
+// Verify op folding for self-mapped pairs
+module attributes {} {
+  // CHECK-LABEL: collective_permute_self_mapped_pair
+  func.func public @collective_permute_self_mapped_pair(%arg0: tensor<1x1x8192x512xf32>) -> (tensor<1x1x8192x512xf32> {jax.result_info = ""}) {
+    %0 = tensor.empty() : tensor<1x1x8192x512xf32>
+    %1 = "ttir.collective_permute"(%arg0, %0) <{source_target_pairs = dense<[[0, 0], [1, 2]]> : tensor<2x2xi64>}> : (tensor<1x1x8192x512xf32>, tensor<1x1x8192x512xf32>) -> tensor<1x1x8192x512xf32>
+    // CHECK: "ttnn.collective_permute"
+    // CHECK-SAME: source_target_pairs = dense<{{([[])}}[1, 2]]> : tensor<1x2xi64>
+    // CHECK-NOT: [0, 0]
```
Comment (azecevicTT) [inline test/ttmlir/Dialect/TTNN/ccl/collective_permute/collective_permute_positive.mlir:38]: This is redundant, because if previous like (`CHECK-SAME`) succeeds FileCheck doesn't go back, and `[0, 0]` can only really appear in `source_target_pairs`.

File: test/ttmlir/Dialect/TTNN/ccl/collective_permute/collective_permute_positive.mlir [hunk 6]
```diff
@@ -1,14 +1,54 @@
...
-// -----
-
 // Verify lowering of ttir collective_permute to ttnn ops
 module attributes {} {
-  func.func public @collective_permute_invalid_duplicate_sources(%arg0: tensor<1x1x8192x512xf32>) -> (tensor<1x1x8192x512xf32> {jax.result_info = ""}) {
+  // CHECK-LABEL: collective_permute_valid_case
+  func.func public @collective_permute_valid_case(%arg0: tensor<1x1x8192x512xf32>) -> (tensor<1x1x8192x512xf32> {jax.result_info = ""}) {
     %0 = tensor.empty() : tensor<1x1x8192x512xf32>
     %1 = "ttir.collective_permute"(%arg0, %0) <{source_target_pairs = dense<[[0, 1], [1, 0]]> : tensor<2x2xi64>}> : (tensor<1x1x8192x512xf32>, tensor<1x1x8192x512xf32>) -> tensor<1x1x8192x512xf32>
+    // CHECK: "ttnn.collective_permute"
```
Comment (azecevicTT) [inline test/ttmlir/Dialect/TTNN/ccl/collective_permute/collective_permute_positive.mlir:10]: Can you add `CHECK-SAME: source_target_pairs = (expected result)` here just to verify the expectation that fold didn't change anything in this case.

File: lib/Dialect/TTNN/IR/TTNNOps.cpp (hunk 1 above)
Comment (hkwonTT) [reply lib/Dialect/TTNN/IR/TTNNOps.cpp]: updated as you suggested.

File: lib/Dialect/TTNN/IR/TTNNOps.cpp (hunk 4 above)
Comment (hkwonTT) [reply lib/Dialect/TTNN/IR/TTNNOps.cpp]: Is this suggestion for better readability? Anyway, I updated code as you suggested.

File: test/ttmlir/Dialect/TTNN/ccl/collective_permute/collective_permute_positive.mlir (hunk 5 above)
Comment (hkwonTT) [reply test/ttmlir/Dialect/TTNN/ccl/collective_permute/collective_permute_positive.mlir]: Yes, you are right. I tried to make cross-check, but this redundancy can make unexpected false alarm in the future. I removed it.

File: test/ttmlir/Dialect/TTNN/ccl/collective_permute/collective_permute_positive.mlir (hunk 6 above)
Comment (hkwonTT) [reply test/ttmlir/Dialect/TTNN/ccl/collective_permute/collective_permute_positive.mlir:10]: Sure. We shouldd check it. Thanks.

**Review by tapspatel — APPROVED** [review]
looks good!

File: lib/Dialect/TTNN/IR/TTNNOps.cpp (hunk 4 above)
Comment (azecevicTT) [reply lib/Dialect/TTNN/IR/TTNNOps.cpp]: > Is this suggestion for better readability?

I would say yes, having indices vs iterators and it's more idiomatic loop.

**Review by azecevicTT — APPROVED** [review]
Thanks for addressing comments.
